### ISSUE TITLE
Repository and module layout

### ISSUE THREAD

**Issue by danielgentile22** [danielgentile22/chess-engine#10]
Part of #1

## Question

Decide the repository shape: engine crate structure (board, movegen, search, eval, uci), the Python trainer, the test tooling, and shared data directories. Apply /architect and /codebase-design for deep-module boundaries. Answer with a directory tree and the interface between engine and trainer.

**Comment by danielgentile22** [comment]
## Answer

**One repository, a two-crate cargo workspace, flat modules with named split points, and a file-on-disk interface between engine and trainer.**

### Scope of the repository

This repo holds the engine, its data generator, its texel tuner, the `bullet` trainer setup, and the test tooling. The queue-and-dashboard service stays out: it is a long-running web service with a database and its own deploy story, and #6 may conclude that OpenBench is adopted rather than built, which settles it by making it somebody else's repository. Datagen and the tuner cannot be split out, because both link engine internals directly.

### Stack correction

The map's Notes said "Python plus PyTorch trainer". That is superseded. #4 picked `bullet`, which is Rust and MIT. The layout is written for `bullet`, with no Python scaffolding reserved for the `nnue-pytorch` fallback: if #16 finds Metal broken, the fallback is a change of language, data handling and export step, not a directory rename, and a placeholder directory would not soften it.

Python survives in this repo only in `tools/`, for match running and the statistical gate.

### Workspace

Two crates, split at the dependency boundary so that `cargo build -p engine` never compiles Metal. `bullet` links `objc2-metal` and Metal Performance Shaders, a GPU dependency tree three months old on this platform; sharing a crate with the engine would let a broken `bullet` update block engine work.

```
chess-engine/
  Cargo.toml                  # workspace: crates/engine, crates/trainer
  rust-toolchain.toml
  LICENSE                     # MIT
  README.md
  CLAUDE.md
  .github/workflows/ci.yml    # build, test, clippy, fmt --check
  crates/
    engine/
      Cargo.toml              # [features] datagen, tuner
      src/
        lib.rs
        board.rs              # bitboards, position, make/unmake, zobrist
        movegen.rs
        tt.rs                 # transposition table: own module from day one
        search.rs
        eval.rs
        uci.rs
        bin/
          chess-engine.rs     # the UCI binary; name provisional
          datagen.rs          # required-features = ["datagen"]
          tuner.rs            # required-features = ["tuner"]
      tests/
        perft.rs              # runs under plain `cargo test`
    trainer/
      Cargo.toml              # bullet, features = ["metal"]
      src/main.rs             # network shape and hyperparameters
  networks/
    *.nnue                    # committed to git, roughly 197 KB each
  tools/
    requirements.txt
    *.py                      # match running, pentanomial SPRT gate
  docs/
    research/
```

Datagen and the tuner live inside the engine crate, behind features, because both need engine internals: datagen runs real searches to label positions, and the texel loss is defined over quiescence scores, so the tuner calls quiescence search per position. Separate crates would force those internals public, which is the wrong pressure on the engine's interface. akimbo does the same at 294 lines of `src/datagen.rs`.

### Module granularity

Flat files now, not directories. The core is hand-written a few hours a week and a file readable top to bottom is the format this project is optimised for. Directories from day one would mean deciding where `search.rs` divides before knowing how the survey's five tiers land in it.

The transposition table is its own module from the start regardless: shared mutable state with its own layout and replacement policy, whose design the map already lists as an open question.

Split points recorded now so later division is mechanical:

- `search.rs` becomes `search/{alphabeta,quiescence,ordering,history}.rs` at roughly 800 lines.
- `eval.rs` becomes `eval/{classical,nnue}.rs` when the first network lands and both paths coexist, before the classical side is deleted.
- `movegen.rs` becomes `movegen/{magics,generate}.rs` if attack-table setup outgrows the file.

### The engine-to-trainer interface

With a two-crate split, the coupling is a file format, so the format is the interface:

1. `datagen` writes plain text lines, `<FEN> | <score> | <result>`, into `$CHESS_DATA_DIR` (default `~/chess-engine-data`).
2. `bullet-utils convert --from text` produces 32-byte `bulletformat` records.
3. `trainer` trains and exports a quantised network: input layer at 255, output layer at 64, evaluation scale 400.
4. The `.nnue` is committed to `networks/`.
5. The engine embeds it with `include_bytes!` at compile time.

Text rather than binary for the first generation: the likely first bug is a sign error on scores from Black's point of view, which is visible by eye in text and invisible in a 32-byte record. The cost is roughly 7 to 8 GB against 3.2 GB at 100 million positions, plus one conversion pass. Writing `bulletformat` or `viriformat` directly is a named later optimisation, triggered when the cost of regenerating data each network iteration starts to hurt.

Embedding rather than run-time loading, because every change is gated by self-play against the previous version, and the network is part of what makes a version play the way it does. With an external file, "version N versus version N-1" quietly depends on which file each process loaded. An embedded net makes a binary a complete statement of one version, and the match runner needs no per-engine configuration beyond a path. A UCI option for a run-time override gets added when the trainer exists and comparing six nets against one build is a real workflow.

### Artifact storage

Networks are committed to `networks/` as ordinary git objects. At roughly 197 KB they are smaller than several research notes, and committing them means a clone at any commit reproduces exactly the engine that played those games, which is what gating rests on. Git-lfs would add a storage system, a bandwidth quota, and a way for a fresh clone to build a broken binary, to save a few megabytes over years.

Bulk training data lives outside the repository, at `$CHESS_DATA_DIR`, not in a git-ignored `data/`. An ignored directory in the working tree is one `git add -f` away from an 8 GB commit in a public repo's history; an external path cannot have that accident, and it lets the data move to another volume without touching the tree. The cost is one environment variable, documented in the README.

### Testing and continuous integration

Perft is an ordinary Rust integration test in `crates/engine/tests/`, run by `cargo test`. It is a correctness property with constant expected values, and anything that requires remembering a separate command stops being run.

Match running and the sequential probability ratio test gate are Python in `tools/`. Not because Rust could not, but because the reference implementation to read and probably port, OpenBench's `stats.py`, is 173 lines of Python holding both the trinomial and pentanomial tests. Porting statistics across a language boundary while also learning the statistics is two problems where one will do. This is the only Python in the repository.

Continuous integration runs on every push and does four things: `cargo build`, `cargo test`, `cargo clippy`, `cargo fmt --check`. No strength testing in CI. Hosted runners are shared and noisy, and self-play measurement is sensitive enough to timing that a result would reflect the runner rather than the change. CI proves correctness; the Mac Studio proves strength. #18 is about to check whether that machine even schedules concurrent games the way #7 assumes.

### Naming

Directories stay generic (`crates/engine`, `crates/trainer`). The binary target and the UCI `id name` carry the engine's name, which keeps a rename down to two string constants instead of every path, import and script. The name is deferred: the binary is `chess-engine` for now, explicitly provisional.

### What would reopen this

If #16 finds that `bullet` does not train on the M4 Max, the `trainer` crate is replaced by an `nnue-pytorch` setup. That changes the trainer's language, the data handling and the export step, but not the workspace shape, the module layout, or the decision to couple engine and trainer only through files on disk.

Resolves #10.
